File: 2022/day1/day1.hs
import Data.List.Split (splitWhen)
import Data.List (sort)

stringListToIntList :: [String] -> [Int]
stringListToIntList = map read

splitList :: [Char] -> [[Int]]
splitList = map stringListToIntList . splitWhen (=="") . lines

top3 :: [Int] -> Int
top3 = sum . take 3 . reverse . sort

main = do
  content <- readFile "input.txt"

  let sums = map sum $ splitList content

  print "Part 1:"
  print $ maximum sums

  print "Part 2:"
  print $ top3 sums

File: 2022/day1/day1.exs
{:ok, contents } = File.read("input.txt")

sums = contents
  |> String.split("\n")
  |> Enum.chunk_by(fn x -> x != "" end)
  |> Enum.filter(fn x -> x != [""] end)
  |> Enum.map(fn x -> Enum.map(x, fn y -> String.to_integer(y) end) end)
  |> Enum.map(fn x -> Enum.sum(x) end)

# part 1 answer
sums |> Enum.max |> IO.inspect(label: "part 1")

# part 2 answer
sums |> Enum.sort |> Enum.reverse |> Enum.take(3) |> Enum.sum |> IO.inspect(label: "part 2")
